rw_mutex.go

```diff
@@ -124,30 +135,51 @@ func (m *RWMutex) Lock() error {
 
 // Unlock releases the write lock
 func (m *RWMutex) Unlock() error {
-	res := m.collection.FindOneAndUpdate(context.TODO(), bson.M{
-		"lockID": m.lockID,
-		"writer": m.clientID,
-	}, bson.M{
-		"$set": bson.M{
-			"writer": "",
+
+	deleteRes, err := m.collection.DeleteOne(
+		context.TODO(),
+		bson.M{
+			"lockID": m.lockID,
+			"writer": m.clientID,
+			"$or":    emptyReaderQuery["$or"],
 		},
-	})
-	if res.Err() == mongo.ErrNoDocuments {
-		return ErrNotOwner
+	)
+	if err != nil {
+		return err
 	}
-	return res.Err()
-}
 
-// RLock acquires the read lock
-func (m *RWMutex) RLock() error {
-	lock, err := m.findOrCreateLock()
+	if deleteRes.DeletedCount > 0 {
+		return nil
+	}
+
+	updateRes, err := m.collection.UpdateOne(
```

Comment:
I'm confused by this case. The difference in the query btw this and the above DeleteOne is just the filter on the emptyReadyQuery, and there should always never be any readers when the write lock was acquired. We should never have a case where the above DeleteOne query didn't match anything but this UpdateOne query does.